sky/shell/platform/ios/platform_view_ios.mm

```diff
@@ -299,9 +380,15 @@ bool UpdateStorageSizeIfNecessary() {
 void PlatformViewIOS::RunFromSource(const std::string& main,
                                     const std::string& packages,
                                     const std::string& assets_directory) {
-  // TODO(johnmccutchan): Call to the iOS UI thread so that services work
-  // properly like we do in PlatformViewAndroid.
-  engine().RunFromSource(main, packages, assets_directory);
+  auto latch = new ftl::ManualResetWaitableEvent();
+
+  dispatch_async(dispatch_get_main_queue(), ^{
+    SetupAndLoadFromSource(main, packages, assets_directory);
+    latch->Signal();
+  });
+
+  latch->Wait();
+  delete latch;
```

Comment:
> Why not just allocate the latch on the stack?

Was trying to work around Objective-C block capture semantics when it comes C++ objects. I suppose I can get a hold of the platform thread task runner somehow and use only `base::` or `ftl::` APIs.

> I'm losing track of which threads are allowed to wait on which other threads.

This patch just applies the iOS/Mac updates necessary for hot reloading as the ones applied for Android at https://github.com/flutter/engine/pull/2944. I will try to document that somehow. With regards to other cases of threads waiting on other threads, I don't recall introducing unconditional waits in any other instance recently. I'd be happy to document all such instance but they should be extremely rare and be required for specific use cases like hot reloading. Can you point me to other cases where you thought we haven't made the threading rules clear?